util/oci/client.go

```diff
@@ -0,0 +1,361 @@
+	ociRepo := strings.TrimPrefix(repoURL, "oci://")
+	repo, err := remote.NewRepository(ociRepo)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize repository: %w", err)
+	}
+
+	repo.PlainHTTP = creds.InsecureHttpOnly
+
+	var tlsConf *tls.Config
+	if !repo.PlainHTTP {
+		tlsConf, err = newTLSConfig(creds)
+		if err != nil {
+			return nil, fmt.Errorf("failed setup tlsConfig: %w", err)
+		}
+	}
+
+	client := &http.Client{Transport: &http.Transport{
+		Proxy:             proxy.GetCallback(proxyUrl, noProxy),
+		TLSClientConfig:   tlsConf,
+		DisableKeepAlives: true,
+	}}
+	repo.Client = &auth.Client{
+		Client: client,
+		Cache:  nil,
+		Credential: auth.StaticCredential(repo.Reference.Registry, auth.Credential{
+			Username: creds.Username,
+			Password: creds.Password,
+		}),
+	}
+
+	c := &nativeOCIClient{
+		creds:    creds,
+		repoURL:  ociRepo,
+		proxy:    proxyUrl,
+		repoLock: repoLock,
+		repo:     repo,
+	}
+	for i := range opts {
+		opts[i](c)
+	}
+	return c, nil
+}
+
+// nativeOCIClient implements Client interface using oras-go
+type nativeOCIClient struct {
+	creds          Creds
+	repoURL        string
+	proxy          string
+	repo           *remote.Repository
+	repoLock       sync.KeyLock
+	indexCache     indexCache
+	repoCachePaths argoio.TempPaths
+}
+
+func (c *nativeOCIClient) TestRepo(ctx context.Context) (bool, error) {
+	err := c.repo.Tags(ctx, "", func(tags []string) error {
+		return nil
+	})
+	return err == nil, err
+}
+
+func (c *nativeOCIClient) Extract(ctx context.Context, digest string, project string, manifestMaxExtractedSize int64, disableManifestMaxExtractedSize bool) (string, argoio.Closer, error) {
+	cachedPath, err := c.getCachedPath(digest, project)
+	if err != nil {
+		return "", nil, err
+	}
+
+	c.repoLock.Lock(cachedPath)
+	defer c.repoLock.Unlock(cachedPath)
+
+	exists, err := fileExist(cachedPath)
+	if err != nil {
+		return "", nil, err
+	}
+
+	if !exists {
+		// Create scratch dir for the OCI store. This will be put into a TAR archive.
+		tempDir, err := files.CreateTempDir(os.TempDir())
+		if err != nil {
+			return "", nil, err
+		}
+		defer os.RemoveAll(tempDir)
+
+		store := newTarFileStore(tempDir, cachedPath)
+
+		// Copy remote repo at the given digest to the scratch dir.
+		_, err = oras.Copy(ctx, c.repo, digest, store, digest, oras.DefaultCopyOptions)
+		if err != nil {
+			return "", nil, err
+		}
+	}
+
+	manifestsDir, err := files.CreateTempDir(os.TempDir())
+	if err != nil {
+		return manifestsDir, nil, err
+	}
+
+	closer, err := file.New(manifestsDir)
+	if err != nil {
+		return manifestsDir, closer, err
+	}
+
+	gzippedFile, err := os.Open(cachedPath)
+	if err != nil {
+		return "", nil, err
+	}
+
+	maxSize := manifestMaxExtractedSize
+	if disableManifestMaxExtractedSize {
+		maxSize = math.MaxInt64
+	}
+
+	err = files.Untgz(manifestsDir, gzippedFile, maxSize, false)
+	if err != nil {
+		return "", nil, err
+	}
```

Comment:
According to the [previous discussion](https://github.com/argoproj/argo-cd/pull/18646#discussion_r1654308484), this might not be what we want.

My understanding is that we need something like:

```go
func Extract() {
	// Do something before...

	// 1. On the first download, download the OCI content and save it as a tar file
	ociDir, err := oci.New(cachedPath)
	desc, err := oras.Copy(ctx, repo, "mytag", ociDir, "mytag", oras.DefaultCopyOptions)
	tarball, err := tarDirectory(cachedPath)

	// 2. Afterwards (and for all subsequent extractions of a given digest), initialize a new OCI store using oci.NewFromTar from cachedPath.
	readonlyOCI, err = oci.NewFromTar(cachedPath)

	// 3. Extract the manifest content from the local OCI store to a temp dir. 
	tempDir := getTempDir()
	fs, err := customFileStore.New(tempDir)
	// copy the whole artifacts
	desc, err := oras.Copy(ctx, readonlyOCI, "mytag", fs, "mytag", oras.DefaultCopyOptions)

	// Do something after...
}

type CustomFileStore struct {
	workingDir string
	file.Store
}

func (cfs *CustomFileStore) Push(ctx context.Context, desc ocispec.Descriptor, content io.Reader) error {
	// If the descriptor has no filename, assign one by adding the "org.opencontainers.image.title" annotation
	if desc.Annotations["org.opencontainers.image.title"] == "" {
		desc.Annotations["org.opencontainers.image.title"] =  generateFileName(desc) // somehow generate a filename
	}

	// Let file store to save the content to the specified filename
	cfs.Store.Push(ctx, desc, content)

	// Now, the content has been saved to the specified filename
	isORASPacked := desc.Annotations[AnnotationUnpack] == "true"
	if !isORASPacked && isTarGz(desc.MediaType){
		// If the .tar.gz is packed by ORAS, it should has been unpacked automatically
		// If not, we need to manually extract the .tar.gz file
		filename := desc.Annotations["org.opencontainers.image.title"]
		filepath := getPath(cfs.workingDir, filename)
		extractTarGz(content, desc, filepath)
	}
}
```

Reply:
@Wwwsylvia yeah I think you're right. I initially thought that it'd be enough to just extract the manifest tarball from the image but we will also need to fetch the manifest locally, in order to display the oci metadata.

Reply:
As for generating the file name, I'm not sure if it's so important in the context of Argo CD - the only thing we want to do is to extract the contents of the tarball.

Reply:
Yeah if the file is put in a temp dir and the name doesn't really matter, we can maybe just use a fixed name.